### PR TITLE
planner,expression: Limit projection pushed down scope to avoid potential performance degression

### LINKED ISSUE #53613: Limit projection pushed down scope to avoid potential performance degression
## Enhancement
When projection push down system variable is set on, all logical projections that contain push-downable expressions will generate potential tikv side physical projections. Currently optimizer evaluates the cost of root-side projection and tikv-side projection cost here:
https://github.com/pingcap/tidb/blob/57d0b40cd9610351a6769926e5c9e30ddd5f6b7c/pkg/planner/core/plan_cost_ver2.go#L102
Although it does consider network cost, calculation concurrency, expressions calculation overhead and other factors that may affect performance, in real situation it still have chances to chooses to push down some projections while performance degrades. 
I can't give a decrete example here, but I believe we can't be too careful of such plan change logic. So we can consider adding some strict rules that only projections that satisfy these rules, can be pushed down to tikv side. Currently, I can think out two rules:

1. Projections that only prune columns can be pushed down
2. Projections' top expressions are json search functions or json value attributes functions(https://docs.pingcap.com/tidb/stable/json-functions).

### PR DESCRIPTION
**Pull request by yibin87**
### What problem does this PR solve?

Issue Number: close #53613 

Problem Summary:
Described in #53613 

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/expression/util.go [hunk 1]
```diff
@@ -1450,6 +1450,53 @@ func ContainCorrelatedColumn(exprs []Expression) bool {
...
+				arg0 := v.GetArgs()[0]
+				// Only `->>` which parsed to JSONUnquote(CAST(JSONExtract() AS string)) can be pushed down to tikv
+				if fChild, ok := arg0.(*ScalarFunction); ok {
+					if fChild.FuncName.L != ast.Cast {
+						return false
+					}
+					if fGrand, ok := fChild.GetArgs()[0].(*ScalarFunction); ok {
+						if fGrand.FuncName.L != ast.JSONExtract {
+							return false
+						}
```
Comment (SeaRise) [inline pkg/expression/util.go:1481]:
```suggestion
						}
						continue
```

Comment (SeaRise) [reply pkg/expression/util.go:1481]: It seems like a "continue" is missing.

Comment (yibin87) [reply pkg/expression/util.go:1480]: Indeed not straightforward coding here, and missed some situation. Refact it.

**Review by SeaRise — APPROVED** [review]
lgtm

File: pkg/planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -2767,7 +2767,8 @@ func (p *LogicalProjection) ExhaustPhysicalPlans(prop *property.PhysicalProperty
 		newProps = append(newProps, mppProp)
 	}
 	if newProp.TaskTp != property.CopSingleReadTaskType && ctx.GetSessionVars().AllowProjectionPushDown && p.CanPushToCop(kv.TiKV) &&
-		expression.CanExprsPushDown(pushDownCtx, p.Exprs, kv.TiKV) && !expression.ContainVirtualColumn(p.Exprs) {
+		expression.CanExprsPushDown(pushDownCtx, p.Exprs, kv.TiKV) && !expression.ContainVirtualColumn(p.Exprs) &&
```
[the remark this answers is not in the document]
Comment (yibin87) [reply pkg/planner/core/exhaust_physical_plans.go:2770]: Removed the check from ProjectionBenefitsFromPushedDown

**Review by winoros — APPROVED** [review]

File: pkg/expression/util.go [hunk 3]
```diff
@@ -1450,6 +1450,50 @@ func ContainCorrelatedColumn(exprs []Expression) bool {
...
+		case *ScalarFunction:
+			allColRef = false
+			switch v.FuncName.L {
+			case ast.JSONDepth, ast.JSONLength, ast.JSONType, ast.JSONValid, ast.JSONContains, ast.JSONContainsPath,
+				ast.JSONExtract, ast.JSONKeys, ast.JSONSearch, ast.JSONMemberOf, ast.JSONOverlaps:
+				continue
+			case ast.JSONUnquote:
+				arg0 := v.GetArgs()[0]
+				// Only `->>` which parsed to JSONUnquote(CAST(JSONExtract() AS string)) can be pushed down to tikv
+				if fChild, ok := arg0.(*ScalarFunction); ok {
```
Comment (windtalker) [inline pkg/expression/util.go:1473]: maybe we can use a function like `JSONUnquoteBenefitsFromPushDown` to make the code more clear.

Comment (yibin87) [reply pkg/expression/util.go]: Nice suggestion, updated.

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**